### PR TITLE
Add debug endpoint to get minion task status.

### PR DESCRIPTION
**Pull request by ramabme**
## Description
Given a taskType, fetches information for all the HelixJobs for the taskType.
For each of the HelixJobs, collects status of the (sub)tasks in the taskbatch.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

**Comment by ramabme** [comment]
Add an initial implementation of debug endpoint to get information about the minion PinotTasks queued and information about the subtasks running under each PinotTask batch. 
The endpoint has a verbositylevel. By default we print only high level summary for completed tasks. For incomplete subtasks prints more detail. If verbositylevel is set to >=10 , prints for all tasks.

An example output with default verbosity - the first and the third PinotTask are completed tasks with only high level info ,

{
  "Task_RealtimeToOfflineSegmentsTask_1628833546894": {
    "subTaskInfos": null,
    "subTaskCount": {
      "waiting": 0,
      "error": 0,
      "total": 1,
      "running": 0,
      "completed": 1
    },
    "startTime": "2021-08-12 22:46:12 PDT",
    "taskState": "COMPLETED",
    "executionStartTime": "2021-08-12 22:46:12 PDT"
  },
  "Task_RealtimeToOfflineSegmentsTask_1628833651118": {
    "subTaskInfos": [
      {
        "startTime": "2021-08-12 22:47:57 PDT",
        "info": "",
        "taskConfig": {
          "configs": {
            "uploadURL": "http://localhost:18998/segments",
            "downloadURL": "http://localhost:18998/segments/mytable/mytable__0__5__20210813T0544Z,http://localhost:18998/segments/mytable/mytable__1__5__20210813T0544Z",
            "windowStartMs": "1388707200000",
            "segmentName": "mytable__0__5__20210813T0544Z,mytable__1__5__20210813T0544Z",
            "windowEndMs": "1388793600000",
            "tableName": "mytable_REALTIME"
          },
          "taskType": "RealtimeToOfflineSegmentsTask"
        },
        "state": "RUNNING",
        "participant": "Minion_192.168.1.21_9514",
        "finishTime": null,
        "taskId": "Task_RealtimeToOfflineSegmentsTask_1628833651118_0"
      }
    ],
    "subTaskCount": {
      "waiting": 0,
      "error": 0,
      "total": 1,
      "running": 1,
      "completed": 0
    },
    "startTime": "2021-08-12 22:47:57 PDT",
    "taskState": "IN_PROGRESS",
    "executionStartTime": "2021-08-12 22:47:57 PDT"
  },
  "Task_RealtimeToOfflineSegmentsTask_1628833502672": {
    "subTaskInfos": null,
    "subTaskCount": {
      "waiting": 0,
      "error": 0,
      "total": 1,
      "running": 0,
      "completed": 1
    },
    "startTime": "2021-08-12 22:45:14 PDT",
    "taskState": "COMPLETED",
    "executionStartTime": "2021-08-12 22:45:14 PDT"
  }
}

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 1]
```diff
@@ -452,6 +640,25 @@ public void addToError(int error) {
...
+      if (state == null) {
+        // task is not yet assigned to a participant
+        addToWaiting(1);
+      } else if (state.equals(TaskPartitionState.INIT) || state.equals(TaskPartitionState.RUNNING)) {
+        addToRunning(1);
+      } else if (state.equals(TaskPartitionState.TASK_ERROR)) {
+        addToError(1);
+      } else if (state.equals(TaskPartitionState.COMPLETED)) {
+        addToCompleted(1);
+      }
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:659]: Let's add a catch all state (e.g. `Unknown`) in case we got other states

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 2]
```diff
@@ -452,6 +640,25 @@ public void addToError(int error) {
       _error += error;
     }
 
+    public void addToCompleted(int completed) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:643]: We can remove all these `addTo*` and only allow `addTaskState`

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 3]
```diff
@@ -427,10 +491,134 @@ private static String getTaskType(String name) {
...
+    public String getInfo() {
+      return _info;
+    }
+
+    public PinotTaskConfig getTaskConfig() {
+      return _taskConfig;
+    }
+  }
+
   public static class TaskCount {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:617]: Let's fix the order of these fields by putting a `JsonPropertyOrder` annotation

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 4]
```diff
@@ -427,10 +491,134 @@ private static String getTaskType(String name) {
     return name.split(TASK_NAME_SEPARATOR)[1];
   }
 
+  public static class TaskDebugInfo {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:494]: Let's fix the order of these fields by putting a `JsonPropertyOrder` annotation

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 5]
```diff
@@ -427,10 +491,134 @@ private static String getTaskType(String name) {
...
+      return _subTaskCounts;
+    }
+
+    public List<TaskPartitionDebugInfo> getSubTaskInfos() {
+      return _subTaskInfos;
+    }
+  }
+
+  public static class TaskPartitionDebugInfo {
+    private String _taskId = null;
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:549]: (nit) no need to explicitly put `= null` for member variables

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 6]
```diff
@@ -381,6 +377,74 @@ public synchronized TaskState getTaskState(String taskName) {
...
+   * Given a taskType, helper method to debug all the HelixJobs for the taskType.
+   * For each of the HelixJobs, collects status of the (sub)tasks in the taskbatch.
+   *
+   * @param taskType      Pinot taskType / Helix JobQueue
+   * @param verboseLevel  By default, does not show details for completed tasks.
+   *                      If verboseLevel >= 10, shows details for all tasks.
+   * @return Map of Pinot Task Name to TaskDebugInfo. TaskDebugInfo contains details for subtasks.
+   */
+  public synchronized Map<String, TaskDebugInfo> getTaskDebugInfo(String taskType, int verboseLevel) {
+    Map<String, TaskDebugInfo> taskDebugInfos = new HashMap<>();
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:390]: Use a `TreeMap` so that the entries are ordered by the task name

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 7]
```diff
@@ -381,6 +377,74 @@ public synchronized TaskState getTaskState(String taskName) {
...
+    WorkflowContext workflowContext = _taskDriver.getWorkflowContext(getHelixJobQueueName(taskType));
+    if (workflowContext == null) {
+      return taskDebugInfos;
+    }
+    boolean showCompleted = verboseLevel >= 10;
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:395]: Why do we choose >= 10 instead of > 1?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java [hunk 8]
```diff
@@ -131,6 +131,15 @@ public StringResultResponse getTaskQueueStateDeprecated(
     return _pinotHelixTaskResourceManager.getTaskStatesByTable(taskType, tableNameWithType);
   }
 
+  @GET
+  @Path("/tasks/{taskType}/debug")
+  @ApiOperation("Fetch information for all the tasks for the given task type")
+  public Map<String, PinotHelixTaskResourceManager.TaskDebugInfo> getTaskDebugInfo(
+      @ApiParam(value = "Task type", required = true) @PathParam("taskType") String taskType,
+      @ApiParam(value = "verbosity (default prints for running and error tasks. Value of >=10 prints for all tasks)") @DefaultValue("1") @QueryParam("verboseLevel") int verboseLevel) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTaskRestletResource.java:139]: Let's use `verbosity` to be consistent with the segment debug info

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 9]
```diff
@@ -381,6 +377,74 @@ public synchronized TaskState getTaskState(String taskName) {
...
+    long timeMillis;
+    SimpleDateFormat SIMPLE_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss z", Locale.getDefault());
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:397]: We should make this a util to avoid different class using different format. We can do it in a separate PR

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 10]
```diff
@@ -381,6 +377,74 @@ public synchronized TaskState getTaskState(String taskName) {
...
+    SIMPLE_DATE_FORMAT.setTimeZone(TimeZone.getDefault());
+    Map<String, TaskState> helixJobStates = workflowContext.getJobStates();
+    for (String helixJobName : helixJobStates.keySet()) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:400]: Use `entrySet()` to avoid another lookup to get the job state

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 11]
```diff
@@ -381,6 +377,74 @@ public synchronized TaskState getTaskState(String taskName) {
...
+   * @return Map of Pinot Task Name to TaskDebugInfo. TaskDebugInfo contains details for subtasks.
+   */
+  public synchronized Map<String, TaskDebugInfo> getTaskDebugInfo(String taskType, int verboseLevel) {
+    Map<String, TaskDebugInfo> taskDebugInfos = new HashMap<>();
+    WorkflowContext workflowContext = _taskDriver.getWorkflowContext(getHelixJobQueueName(taskType));
+    if (workflowContext == null) {
+      return taskDebugInfos;
+    }
+    boolean showCompleted = verboseLevel >= 10;
+    long timeMillis;
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:396]: (nit) Don't define it here

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 12]
```diff
@@ -381,6 +377,74 @@ public synchronized TaskState getTaskState(String taskName) {
...
+    SimpleDateFormat SIMPLE_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss z", Locale.getDefault());
+    SIMPLE_DATE_FORMAT.setTimeZone(TimeZone.getDefault());
+    Map<String, TaskState> helixJobStates = workflowContext.getJobStates();
+    for (String helixJobName : helixJobStates.keySet()) {
+      String pinotTaskName = getPinotTaskName(helixJobName);
+      TaskDebugInfo taskDebugInfo = new TaskDebugInfo();
+      taskDebugInfo.setTaskState(workflowContext.getJobState(helixJobName));
+      timeMillis = workflowContext.getJobStartTime(helixJobName);
+      taskDebugInfo.setStartTime((timeMillis <= 0) ? null : SIMPLE_DATE_FORMAT.format(timeMillis));
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:405]: (nit) for readability. Same for other time handling
```suggestion
      long jobStartTimeMs = workflowContext.getJobStartTime(helixJobName);
      if (jobStartTimeMs > 0) {
        taskDebugInfo.setStartTime(SIMPLE_DATE_FORMAT.format(jobStartTimeMs));
      }
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 13]
```diff
@@ -381,6 +377,74 @@ public synchronized TaskState getTaskState(String taskName) {
...
+        for (int partition : partitionSet) {
+          // First get the partition's state and update the subTaskCount
+          TaskPartitionState partitionState = jobContext.getPartitionState(partition);
+          subTaskCount.addTaskState(partitionState);
+          if (!showCompleted) {
+            // Skip details for COMPLETED tasks
+            if (partitionState != null && partitionState.equals(TaskPartitionState.COMPLETED)) {
+              continue;
+            }
+          }
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:423]:
```suggestion
          // Skip details for COMPLETED tasks
          if (!showCompleted && partitionState == TaskPartitionState.COMPLETED) {
            continue;
          }
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 14]
```diff
@@ -381,6 +377,74 @@ public synchronized TaskState getTaskState(String taskName) {
...
+      taskDebugInfo.setStartTime((timeMillis <= 0) ? null : SIMPLE_DATE_FORMAT.format(timeMillis));
+      JobContext jobContext = _taskDriver.getJobContext(helixJobName);
+      if (jobContext != null) {
+        JobConfig jobConfig = _taskDriver.getJobConfig(helixJobName);
+        timeMillis = jobContext.getExecutionStartTime();
+        taskDebugInfo.setExecutionStartTime((timeMillis <= 0) ? null : SIMPLE_DATE_FORMAT.format(timeMillis));
+        Set<Integer> partitionSet = jobContext.getPartitionSet();
+        TaskCount subTaskCount = new TaskCount();
+        subTaskCount.addToTotal(partitionSet.size());
+        for (int partition : partitionSet) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:414]: (Optional) We can avoid a lot of redundant map lookups by directly accessing the `jobContext.getRecord().getMapFields().entrySet()`. You may read the `JobContext` code to understand how each field is extracted

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 7 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:395]: thought we might add other verbose-levels. So made this a high level. 
If u do not foresee this, will make it > 1.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 14 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:414]: The enum for the keys of the mapField is private to the JobContext class. 
```
private enum ContextProperties {
    START_TIME, // Time at which this JobContext was created
    STATE,
    .......
    .......
  }
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 1 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:655]: added

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 2 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java]: removed the addTo* and only setter is addTaskState.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 3 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:629]: done

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 4 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:503]: done

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 5 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java]: cleaned.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 10 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java]: cleaned

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 11 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java]: cleaned

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 12 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java]: cleaned

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 13 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java]: cleaned

**Comment by ramabme** [comment]
Cleaned up based on the review.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 15]
```diff
@@ -427,29 +498,161 @@ private static String getTaskType(String name) {
...
+    }
+
+    /* Update count based on state for each task running under a HelixJob/PinotTask */
+    public void addTaskState(TaskPartitionState state) {
+      _total++;
+      // Helix returns state as null if the task is not enqueued anywhere yet
+      if (state == null) {
+        // task is not yet assigned to a participant
+        _waiting++;
+      } else if (state.equals(TaskPartitionState.INIT) || state.equals(TaskPartitionState.RUNNING)) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:647]: (nit) let's use `switch` here

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 16]
```diff
@@ -427,29 +498,161 @@ private static String getTaskType(String name) {
...
-    private int _error;     // Number of tasks in error
-    private int _running;   // Number of tasks currently running in minions
-    private int _total;     // Total number of tasks in the batch
+  @JsonPropertyOrder({"taskState", "subTaskCount", "startTime", "executionStartTime", "subTaskInfos"})
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  public static class TaskDebugInfo {
+    private String _startTime;
+    private String _executionStartTime;
+    private TaskState _taskState;
+    private TaskCount _subTaskCount;
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:507]: (nit) I think `subtask` is a word, so name the variable `_subtaskCount` and `_subtaskInfos`? Same for the methods

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java [hunk 17]
```diff
@@ -427,29 +498,161 @@ private static String getTaskType(String name) {
...
-    public void addToRunning(int running) {
-      _running += running;
+    public List<TaskPartitionDebugInfo> getSubTaskInfos() {
+      return _subTaskInfos;
+    }
+  }
+
+  @JsonPropertyOrder({"taskId", "state", "startTime", "finishTime", "participant", "info", "taskConfig"})
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  public static class TaskPartitionDebugInfo {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java:559]: Suggest renaming to `SubtaskDebugInfo` to match the json field name?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM with minor comments

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 15 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java]: done

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 16 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java]: changed subTask/SubTask to subtask/Subtask

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java (hunk 17 above)
Comment (ramabme) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotHelixTaskResourceManager.java]: Renamed TaskPartitionDebugInfo to SubtaskDebugInfo

**Comment by kishoreg** [comment]
Why do we call this debug endpoint? Getting state of a task should be a first class api rt

**Comment by ramabme** [comment]
> Why do we call this debug endpoint? Getting state of a task should be a first class api rt

I guess the original motivation was to get information/errors/msgs from Helix to debug failing or long pending tasks. And to figure out and show selective but sufficient information for the tasks of a taskType and for each task(batch) information about subtasks. Jackie/Mayank can correct me.
 
There is an existing API that only gives the state(single value) about the Task(batch) but no details .

**Comment by kishoreg** [comment]
> I guess the original motivation was to get information/errors/msgs from Helix to debug failing or long pending tasks. And to figure out and show selective but sufficient information for the tasks of a taskType and for each task(batch) information about subtasks. Jackie/Mayank can correct me.
> 
> There is an existing API that only gives the state(single value) about the Task(batch) but no details

What is the expected workflow from user
-

1. Get all the tasks and then get debug for a specific task id OR
2. Get all debug info for all tasks

As a developer, I can see the need for 1. Not sure why I would do 2.

**Comment by Jackie-Jiang** [comment]
@kishoreg
We plan to provide the following APIs:
1. Debug info for a given task type to get a global view for all the tasks under this type (in this PR)
2. Debug info for a given task type and id (single task)
3. Get the subtask state count for a given task type

Then we can hook them into the cluster manager UI
